File: leetcode/merge-k-sorted-lists/merge-k-sorted-lists-dc.java
// LeetCode #23 - Merge k Sorted Lists
// https://leetcode.com/problems/merge-k-sorted-lists/
// Using recursive divide and conquer.

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        return mergeLists(lists, 0, lists.length);
    }

    private ListNode mergeLists(ListNode[] lists, int from, int toExclusive) {
        var count = toExclusive - from;

        switch (count) {
        case 0:
            return null;

        case 1:
            return lists[from];

        default:
            var mid = from + count / 2;

            return merge(mergeLists(lists, from, mid),
                         mergeLists(lists, mid, toExclusive));
        }
    }

    private ListNode merge(ListNode first, ListNode second) {
        var sentinel = new ListNode();
        var pre = sentinel;

        while (first != null && second != null) {
            if (second.val < first.val) {
                pre.next = second;
                second = second.next;
            } else {
                pre.next = first;
                first = first.next;
            }

            pre = pre.next;
        }

        pre.next = (first == null ? second : first);
        return sentinel.next;
    }
}

File: leetcode/merge-k-sorted-lists/merge-k-sorted-lists-dc.cpp
// LeetCode #23 - Merge k Sorted Lists
// https://leetcode.com/problems/merge-k-sorted-lists/
// Using recursive divide and conquer.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    [[nodiscard]]
    static ListNode* mergeKLists(const vector<ListNode*>& lists) noexcept;
};

namespace {
    [[nodiscard]] ListNode* merge(ListNode* left, ListNode* right) noexcept
    {
        auto sentinel = ListNode{};
        auto pre = &sentinel;

        while (left && right) {
            if (right->val < left->val) {
                pre->next = right;
                right = right->next;
            } else {
                pre->next = left;
                left = left->next;
            }

            pre = pre->next;
        }

        pre->next = (left ? left : right);
        return sentinel.next;
    }

    template<typename It>
    [[nodiscard]] ListNode* merge_lists(const It first, const It last) noexcept
    {
        const auto count = distance(first, last);

        switch (count) {
        case 0:
            return nullptr;

        case 1:
            return *first;

        default:
            const auto mid = next(first, count / 2);
            return merge(merge_lists(first, mid), merge_lists(mid, last));
        }
    }
}

ListNode* Solution::mergeKLists(const vector<ListNode*>& lists) noexcept
{
    return merge_lists(cbegin(lists), cend(lists));
}
